diff -urp mongodb-src-r2.2.4.orig/src/SConscript.client mongodb-src-r2.2.4/src/SConscript.client
--- mongodb-src-r2.2.4.orig/src/SConscript.client	2013-03-31 21:41:49.000000000 -0500
+++ mongodb-src-r2.2.4/src/SConscript.client	2013-05-02 15:06:53.145000000 -0500
@@ -99,7 +99,7 @@ for path in clientHeaderDirectories:
 mongoclient_lib = env.Library('mongoclient', clientSource),
 mongoclient_install = env.Install('#/', [
         mongoclient_lib,
-        #env.SharedLibrary('mongoclient', clientSource),
+        env.SharedLibrary('mongoclient', clientSource),
         ])
 env.Alias('mongoclient', mongoclient_install)
 
Only in mongodb-src-r2.2.4/src: SConscript.client.orig
